### ISSUE TITLE
Learning degrades as timescale climbs

### ISSUE THREAD

**Issue by scotthovestadt** [Unity-Technologies/ml-agents#1025]
I'm working on training AI for a simple game similar to Missile Command. The game consists of 2 AIs-- the "attacker" and the "defender". The attacker shoots missiles down at cities/silos on the ground and the defender shoots at the incoming missiles with his own to defend.

The defender was not training properly and the optimal strategy he could come up with was to just spray missiles randomly around. To understand what was happening, I pulled the defender AI out into his own training course that shot a single missile down from a random position for him to stop. I immediately reset if he missed the single defensive missile I gave him. I figured this would train him quickly to a minimum standard and then I could throw him in with the other brain and they would learn from there.

To understand, you can see a quick video of this training course in action here (timescale=3):
https://www.youtube.com/watch?v=5dtikumqZE4&feature=youtu.be

Unfortunately, this did not work. His score stayed at 0 for >1million steps. I thought that perhaps the delayed reward was causing issues (because the defender first has to go to the right position, lead the incoming missile, shoot, and then 2~ seconds later will get rewarded if he does it correctly) so I then coded up a reward system that, when the defender fired, _IMMEDIATELY_ detected if the missile would stop the incoming single missile. He would reset immediately with/without reward. My plan was to train him on this and then move him to stopping real missiles, which he would then be able to learn.

The training worked very well. The AI was able to fully learn, quickly, all the way up to a perfect score. I then graduated him back to the original course (with real missiles) and it failed utterly. The AI score fell all the way back to zero.

I suspected at this point that there might be a physics issue with actually intercepting the missile. I then verified that the AI will properly train with on a lower timescale (although obviously it takes a long time). I also verified that the training linearly degrades as I turn the timescale up. The AI is capable of learning with actual collisions (at least somewhat) on a timescale up to 50 but it linearly improves all the way down to 1.

The game is built with 2D physics. The missiles move with a constant velocity. I setup the Rigidbody configurations for all moving objects to use "Continuous" collision detection, never sleep, and I also tried turning on Interpolate. (Original settings were just default dynamic Rigidbody.) The missile collisions are detected via a simple 2D collider. These settings had no impact.

One thing I noticed is that the AI trained on a simulation (not using actual collisions, based on my code detecting if he's close to the right area) will end up learning to lead the missile a lot more than necessary to the point where if I try him on a low timescale he gets a score of 0. His learning only works on high timescale!

Here are other things I tried (all had no impact):
- Training on PC and Mac.
- Training against older version of Unity
- `--no-graphics` on and off
- Several combinations of hyperparameters
- Different ways of moving the missiles (velocity, `MovePosition` + interpolate, and `transform.position` with deltaTime).
- More or less environments (agents with single brain). I was using 12 but I tried 2, I tried 20, no real difference.

OK, for anyone that read the wall of text: thanks. Here are my questions for the maintainers:
- Should I expect that for some games I will need to turn the timescale down when training?
- I noticed that the example game with the freezing lasers / banana pickups you guys developed has the timescale set to only 20 but I didn't see it explained anywhere. Did you have a similar problem?
- Do you have any recommendations for things to try?
- What is the impact of having a very low timescale but having a very high number of agents with single brain? Might this be an alternative way to learn faster while keeping timescale low in my case? Do you have a recommendation for which hyperparameters to tweak if I tried this? I assume I want to speed up the gradient descent?

Thanks for your time,
-Scott

**Comment by vincentpierre** [comment]
Wow, the video looks great. This is a super cool project.
The problem you are facing is odd, do you have any logic in Coroutines or in the Update methods ? Can you give more details on the framework you use? You can do training with multiple agents but beyond a certain point, increasing the number of agents will not be helpful anymore. Maybe decreasing the batch_size and buffer_size will help you make more frequent updates of the model when at low speeds but if too small, the network will not train at all. 
The reason the bananas are only at 20 is because it does not change anything to go from 20 to 100 because of hardware limitations.
Are you on v0.4 ? Maybe modify the training configuration in the academy : Set a higher quality setting?
I hope this helps, please keep us updated.

**Comment by scotthovestadt** [comment]
@vincentpierre I do have a coroutine but all it does is wait for the end of frame. All other logic specifically is coded to only use FixedUpdate and fixedDeltaTime-- even the timer, cooldowns of missiles, etc. Perhaps that wait for end of frame is the issue. I'll refactor it out and let you know. Thanks for taking a look.

The control scheme is continuous as well (although now that discrete looks to have branches soon I might have to try that) so my batch_size and buffer_size are pretty high to account for that. I'll play with that after I see if removing all coroutines completely solves the issue.

If you're curious, here's the current state of the attacker v. defender AIs. The defender AI here is pretty stupid because he's learned only by facing the attacker and nothing else:
https://www.youtube.com/watch?v=HkD9-aKHVro&feature=youtu.be

Attacker is doing more or less what I want him to do at this stage without further incentives but the Defender just gets stuck in a local minimum pretty easily. Once I get the curriculum working to pre-warm both AIs before they meet each other I'm hoping for better results.

**Comment by scotthovestadt** [comment]
@vincentpierre Hmm, still not sure what's going on. I eliminated every coroutine in the program and did what I could to verify everything works as 100 timescale. I think it did help to some extent, the AI seems to be able to learn-- just much worse than at timescale 1.

What I've also noticed is that there isn't a hugely significant difference in the rate that steps advance with timescale 1 vs. timescale 100. Sure, timescale 100 seems somewhat faster, but it's definitely not even close to 100x. I'd say it's less than 2x faster in terms of advancing steps (while learning 10x less).

Is it possible that in a situation where an agent might reset VERY OFTEN a high timescale could cause issues? When I'm first training the agent, I need to teach it trigger control, and so it ends up resetting itself _A LOT_. I find just doing a reset works better than a penalty.

Maybe I'll try to put together something very simple to sanity check if this could be an issue.

**Comment by vincentpierre** [comment]
Increasing TimeScale will not work at a certain point because the calculations required to do one step are greater than the time allowed in one step. If the speedup you gain is around x2 then maybe your computations are too complex and x2 timeScale is the best speed you can get. 
Reseting a lot should not be a problem unless the agent resets twice in a row. If the agent is set to Done right after reseting, it will not be able to learn from that experience.

**Comment by scotthovestadt** [comment]
@vincentpierre Thanks for your assist.

I eventually isolated the problem to an issue with my object pool that only triggered in a specific situation. For whatever reason, the training didn't abort and instead it just reported 0 progress forever.

Suggestion: add a command-line option to surface Unity log output when training.

**Comment by vincentpierre** [comment]
The Unity log outputs are copied into the file `unity-environment.log` located nuxt to the executable you used if you are still looking for a way to see the Unity logs. Alternatively, in editor training would allow you to see the Unity logs directly into the unity console
